### DISCUSSION TITLE
Using BFGS Solver

### DISCUSSION THREAD

**Discussion by prajwalka** [idaholab/moose#24798] [Q&A General] [answered]
Is there a way of using BFGS solver instead of the Newton solver for the nonlinear iterations?

**Comment by hugary1995** [comment]
Yes petsc has several quasi newton nonlinear solvers. BFGS is one of them.

**Reply by lindsayad** [reply, marked as answer]
You can use it by simply adding the petsc option `-snes_type qn`

**Reply by prajwalka** [reply]
Thank you very much!

**Reply by tqcthai** [reply]
Hello, 

One more question. If I use the -snes_type_qn, should I specify the preconditioner -pc_type?
  
If I don't specify the -pc_type, it takes nearly 20 iterations for a linear problem, this is nonsense. But using -snes_type_qn, I don't see any difference in the convergent behavior between

petsc_options_iname = '-pc_type  -snes_type_qn'
petsc_options_value = 'lu               lbfgs'

and 

#petsc_options_iname = '-pc_type -pc_hypre_type'
#petsc_options_value = 'lu       boomeramg'

**Reply by GiudGiud** [reply]
Yes there is a difference between the nonlinear solve method (newton, bfgs, PJFNK etc) and the linear solve method (which we partially specify using pc_type, but it s not always that clear especially for Krylov subspace methods). When you specify LU as the preconditioning, the linear solve is short (one step to solve) as the preconditioner inverses the system already. This preconditioner is expensive though. I ll refer you to the PETSc docs for LU as there are some subtletites 

```
#petsc_options_iname = '-pc_type -pc_hypre_type'
#petsc_options_value = 'lu boomeramg'
```
these arent valid. LU is not part of hypre
maybe this?
```
petsc_options_iname = '-pc_type -pc_hypre_type'
petsc_options_value = 'hypre boomeramg'
```

**Reply by tqcthai** [reply]
@GiudGiud thank you. However here I want to study to make a comparison in the convergent behavior between Newton and BFGS.
For the fully Newton iteration, it should be as you mentioned

petsc_options_iname = '-pc_type -pc_hypre_type'
petsc_options_value = 'hypre boomeramg'

However, for quasi-Newton, should specify like this?

petsc_options_iname = '-pc_type   -snes_type_qn'
petsc_options_value =  'hypre         lbfgs'

**Reply by GiudGiud** [reply]
try this
```
petsc_options_iname = '-pc_type -pc_hypre_type -snes_type -snes_qn_type'
petsc_options_value = 'hypre         boomeramg        qn               lbfgs'
```
you really need to check on the PETSc manual to explore these options. They are not standard in MOOSE
https://petsc4py.readthedocs.io/en/stable/manual/snes/#tab-qndefaults

**Reply by tqcthai** [reply]
thank you @GiudGiud, I found out it works for this setup

`  petsc_options_iname = '-pc_type   -pc_hypre_type     -snes_type        -snes_qn_type   -snes_qn_scale_type'
  petsc_options_value = 'lu         boomeramg           qn               lbfgs           jacobian'`

One more thing, do you know how to jump/switch between different solver in Moose Executioner? For instance, using quasi Newton firstly, once the convergence is having a problem switch it into fully Newton.

**Reply by GiudGiud** [reply]
Hello

We will be able to support that once the Executor system is fully deployed
For now you ll have to use Restarting https://mooseframework.inl.gov/application_usage/restart_recover.html

**Reply by lindsayad** [reply]
Note that `-pc_hypre_type boomeramg` is not used for `-pc_type lu`
